File: tools/mapgen/terragen/bridges.py
"""Road water crossings — where a bridge belongs, published by the map (roads R3b).

The road planner prices water rather than forbidding it (`RoadParams.water_
penalty`), so a generated network fords its rivers and inlets: on skerry_reach
four stretches of deck run below the waterline, the longest 992 elmos. Those
stretches are the ONLY places on a map where a bridge span means anything —
scenariogen's `find_crossing` was searching for narrow water inside a region
with no idea where the roads were, and its own docstring made the assumption
out loud ("a bridge is built where the river is thinnest, and that is also
where a road would naturally run"). This module makes that true instead of
assumed: the crossings are read off the planned network, with the chain's
heading taken from the road's own tangent.

WHAT THIS MODULE DOES NOT DO — placement. It publishes crossings into
`mapdata/roads.lua` for the scenario stager to build spans on, and deliberately
does NOT write featureplacer entries, because the map feature path cannot carry
the one number a span over water needs:

  * `CFeatureHandler::LoadFeaturesFromMap` (rts/Sim/Features/FeatureHandler.cpp)
    spawns every map-authored feature at `CGround::GetHeightReal(x, z)` — the
    SEABED under a crossing. The featureplacer config format has no `y` at all
    (rts/Server/FeatureProcessor.cpp parses name/x/z/rot), and
    `floating = true` only zeroes the gravity term (Feature.cpp:533); it does
    not buoy a span up. A map-placed chain therefore lies on the riverbed, which
    is the staircase §M3 measured live (-31.0 / -34.5 / -45.9 / -57.6).
  * `game_scenario.lua`'s `stageFeatures` takes an explicit `y`, and §M3
    measured four spans staged at `y = 0` holding 0.00 across the chain. So the
    scenario path is the one that can lay a level deck, and the map's job is to
    tell it where.

⛔ **A span is DECORATION over a ford, and this is v1 saying so out loud.**
`features/bridges.lua` ships `blocking = false` with a header explaining why:
Spring/Recoil pathing is single-layer, `CFeature::Block()` is all-or-nothing,
and a blocking span is a wall across the gap it exists to open. So units do not
drive ON the deck — they wade the ford UNDER it, which is why
`crossing_is_fordable` is an acceptance rule here rather than a nicety: a
crossing this module publishes is one the map's own vehicles can already cross.
When the `deckHeight` engine ask lands (.tasks/notes/model-integration.md), the
crossings published here are already the right places to raise.

⛔ **AND A MAP CANNOT CLOSE THAT GAP BY SHAPING GROUND** (roads R3c, measured
2026-08-15 — PLAN-maps.md §2j). The obvious terrain answer, "raise a causeway
whose top IS the deck", is arithmetically impossible, not merely expensive:
`CFeature::UpdatePosition` clamps a feature's y UP to `CGround::GetHeightReal`
every tick and never down (Feature.cpp:570), so a span standing on ground `g`
puts its deck at `g + deck_top` (1.5 for `ms_road_bridge`, measured off the
shipped mesh, published as `customparams.deck_top`). Raise the ground under the
crossing by `d` and BOTH the road surface and the deck rise by `d`: the gap is
invariant under every earthwork this module could ever plan. The map's terrain
levers here are real but they are about the ROUTE — how deep the ford is, how
wide, whether the road should be there at all — never about the deck.
"""
from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np

from .roads import resample_by_arclength, sample_height

# Spring's 16-bit heading, and the direction a chain lays along it. Mirrors
# game_scenario.lua's `headingToDir` exactly — heading 0 is FACING_NORTH and a
# feature's local forward at heading 0 is -Z for an RH/glTF-native game, so
# dir = (sin t, -cos t). tests/test_bridges.py reads the cardinals out of
# game_scenario.lua and checks these two against them, so the convention cannot
# drift here without the gadget drifting too.
HEADING_SCALE = 65536.0


def heading_short(dx: float, dz: float) -> int:
    """Spring heading whose chain direction is (dx, dz). 0..65535."""
    theta = math.atan2(dx, -dz)
    return int(round(theta * HEADING_SCALE / (2.0 * math.pi))) % 65536


def heading_dir(short: int) -> tuple[float, float]:
    """Unit chain direction for a heading — the gadget's `headingToDir`."""
    theta = short * (2.0 * math.pi / HEADING_SCALE)
    return math.sin(theta), -math.cos(theta)


# A single 24-elmo segment is a plank and the gap under it is a ditch; past 24
# segments the "gap" is open water and the chain is a landfill. Same window
# scenariogen's blind placer uses (MIN_BRIDGE_SPANS / MAX_BRIDGE_SPANS), stated
# once per placer because they are separate acceptance rules over the same def.
MIN_SPANS = 3
MAX_SPANS = 24

# VEH's wade depth from `moveinfo.tdf`, via terragen.passability.MOVE_CLASSES.
# Imported lazily in `find_crossings` so this module stays importable without
# scipy; the default here is the number, not a second table.
VEH_WADE_DEPTH = 20.0


@dataclass(frozen=True)
class CrossingParams:
    pitch: float = 24.0          # the def publishes it (customparams.chain_pitch)
    min_spans: int = MIN_SPANS
    max_spans: int = MAX_SPANS
    step: float = 8.0            # arclength sampling along the polyline (1 square)
    wade_depth: float = VEH_WADE_DEPTH
    # How far the wet run may bend before a straight chain would walk off it.
    # The afloat test below is the real gate; this reports the shape.
    max_bend_deg: float = 45.0


@dataclass(frozen=True)
class Crossing:
    """A stretch of road under water, sized for a chain of spans."""
    link: int                   # index into net.links
    road_class: int
    x: float                    # CHAIN CENTRE — stageFeatures chains centred
    z: float
    heading: int                # Spring heading short; chain lays along it
    spans: int
    length: float               # wetted deck length along the road (elmos)
    width: float                # bank-to-bank chord the chain covers (elmos)
    max_depth: float            # deepest point of the ford, positive elmos
    bend_deg: float             # heading spread across the run

    def span_centres(self, pitch: float) -> list[tuple[float, float]]:
        """The (x, z) of every segment, exactly as stageFeatures computes them."""
        dx, dz = heading_dir(self.heading)
        out = []
        for i in range(self.spans):
            step = (i - (self.spans - 1) / 2.0) * pitch
            out.append((self.x + dx * step, self.z + dz * step))
        return out


@dataclass(frozen=True)
class Refusal:
    """A wet stretch of road that gets no bridge, and why."""
    link: int
    reason: str
    length: float
    max_depth: float
    x: float
    z: float

    def describe(self) -> str:
        return (f"link {self.link} at ({self.x:.0f}, {self.z:.0f}): "
                f"{self.length:.0f} elmos wet, {self.max_depth:.1f} deep — "
                f"{self.reason}")


def _wet_runs(pts: np.ndarray, h: np.ndarray, water_level: float):
    """Contiguous index ranges of `pts` whose sampled height is under water."""
    wet = h <= water_level
    out = []
    i = 0
    n = len(wet)
    while i < n:
        if wet[i]:
            j = i
            while j + 1 < n and wet[j + 1]:
                j += 1
            out.append((i, j))
            i = j + 1
        else:
            i += 1
    return out


def _arclength_mid(seg: np.ndarray) -> tuple[float, float]:
    d = np.hypot(*(seg[1:] - seg[:-1]).T)
    cum = np.concatenate([[0.0], np.cumsum(d)])
    half = cum[-1] / 2.0
    k = int(np.searchsorted(cum, half))
    k = min(max(k, 1), len(seg) - 1)
    t = (half - cum[k - 1]) / max(d[k - 1], 1e-6)
    p = seg[k - 1] + (seg[k] - seg[k - 1]) * t
    return float(p[0]), float(p[1])


def crossing_is_fordable(max_depth: float, wade_depth: float) -> bool:
    """Can the map's own vehicles cross here without the bridge?

    They have to, because the span is non-blocking decoration — see the module
    header. A wet stretch of deck deeper than VEH's wade is a road that leads
    into water its own traffic cannot follow, and that is a defect in the ROUTE,
    not a bridge opportunity: reported as a refusal so it is visible rather than
    quietly decorated over.
    """
    return max_depth <= wade_depth


def find_crossings(net, height: np.ndarray, cellsize: float,
                   water_level: float = 0.0,
                   p: CrossingParams | None = None,
                   ) -> tuple[list[Crossing], list[Refusal]]:
    """Every stretch of `net` under water, graded into crossings and refusals.

    `height` is the DELIVERED surface — call this after `flatten_network`, or
    the crossing is measured on ground the road no longer stands on.

    No RNG: runs are walked in link order, so two runs on one seed cannot
    disagree about where a river is.
    """
    p = p or CrossingParams()
    crossings: list[Crossing] = []
    refusals: list[Refusal] = []

    for idx, ln in enumerate(net.links):
        pts = resample_by_arclength(ln.polyline, p.step)
        if len(pts) < 2:
            continue
        hh = sample_height(height, pts, cellsize)
        for a, b in _wet_runs(pts, hh, water_level):
            seg = pts[a:b + 1]
            if len(seg) < 2:
                continue
            length = float(np.sum(np.hypot(*(seg[1:] - seg[:-1]).T)))
            depth = float(water_level - hh[a:b + 1].min())
            cx, cz = _arclength_mid(seg)
            chord = seg[-1] - seg[0]
            width = float(np.hypot(*chord))
            heading = heading_short(float(chord[0]), float(chord[1]))
            # How far the run bends: the widest angle between any sample's
            # local tangent and the chord the chain will lay along.
            tang = seg[1:] - seg[:-1]
            keep = np.hypot(tang[:, 0], tang[:, 1]) > 1e-6
            bend = 0.0
            if keep.any() and width > 1e-6:
                u = chord / width
                cosang = np.clip(
                    (tang[keep] @ u) / np.hypot(tang[keep, 0], tang[keep, 1]),
                    -1.0, 1.0)
                bend = float(np.degrees(np.arccos(cosang)).max())

            def refuse(reason):
                refusals.append(Refusal(idx, reason, length, depth, cx, cz))

            if not crossing_is_fordable(depth, p.wade_depth):
                refuse(f"deeper than VEH wades ({p.wade_depth:.0f}) — the "
                       f"ROUTE is broken here, not the decoration")
                continue
            # Sized by FLOOR over the water, never ceil, and measured on the
            # chord the chain actually lays along rather than on the wandering
            # deck length: segment i sits at (i - (n-1)/2) * pitch from the
            # centre, so n = floor(water / pitch) keeps the outermost centre
            # strictly inside the water. §M4 learned this the hard way — a deck
            # sized to reach the banks put half its chain on rising ground,
            # where `floating` does not defeat the ground clamp.
            water = width - 2.0 * cellsize
            spans = int(water // p.pitch) if water > 0 else 0
            if spans < p.min_spans:
                refuse(f"{spans} spans — a plank over a ditch")
                continue
            if spans > p.max_spans:
                refuse(f"{spans} spans — open water, not a river; a "
                       f"{p.max_spans}-span limit makes this a causeway")
                continue
            cand = Crossing(idx, ln.road_class, cx, cz, heading, spans,
                            length, water, depth, bend)
            # EVERY SPAN CENTRE MUST BE OVER WATER — asked of the same
            # arithmetic stageFeatures will use, not of the floor() above.
            centres = np.asarray(cand.span_centres(p.pitch))
            if float(sample_height(height, centres, cellsize).max()) > water_level:
                refuse(f"chain of {spans} leaves the water (bend "
                       f"{bend:.0f} deg over the run)")
                continue
            crossings.append(cand)
    return crossings, refusals


def emit_crossings_lua(crossings: list[Crossing], defname: str = "ms_road_bridge",
                       indent: str = "        ") -> list[str]:
    """`crossings = { ... }` rows for mapdata/roads.lua.

    `y` is NOT emitted: the waterline is the stager's business (it stages spans
    at y = 0 and the def floats), and a height baked into map data would be a
    second copy of the map's water plane.
    """
    out = []
    for c in crossings:
        out.append(
            f"{indent}{{ def = \"{defname}\", x = {c.x:.0f}, z = {c.z:.0f}, "
            f"heading = {c.heading}, spans = {c.spans}, "
            f"class = {c.road_class}, width = {c.width:.0f}, "
            f"depth = {c.max_depth:.1f} }},")
    return out

File: data/games/metalstorm/features/bridges.lua
-- features/bridges.lua — tileable crossing spans (PLAN-metalstorm-model-integration §M3).
--
-- See features/README.md for the load path and the key table.
--
-- ============================================================================
-- WHY THESE ARE NON-BLOCKING (the §M3 "pathable on top" ask, answered)
-- ============================================================================
-- M3 asks for bridges that units path across ON TOP of while the span blocks
-- underneath. **The engine cannot express that, and the gap is not in these
-- defs.** Spring/Recoil pathing is single-layer: `CFeature::Block()`
-- (rts/Sim/Features/Feature.cpp:237) either stamps the footprint into the
-- ground-blocking map or it does not. There is no deck height, no over/under,
-- no multi-level nav mesh. The one permeability mechanism this fork added —
-- `FootprintProfileHandler`'s `underpass` move-class lists
-- (rts/Sim/MoveTypes/FootprintProfile.cpp) — hangs off `UnitDef`
-- (rts/Sim/Units/UnitDef.h:266) and is never consulted for a feature.
--
-- So the choice is binary, and `blocking = false` is the right half of it:
--
--   blocking = true   a bridge is a WALL across the gap it exists to open.
--                     Every unit routes around the crossing. This is strictly
--                     worse than no bridge at all.
--   blocking = false  a cosmetic span. Units cross wherever the TERRAIN is
--                     already passable, and the bridge is the visual + narrative
--                     explanation of the crossing. Chokepoint gameplay comes
--                     from the terrain the bridge is placed on, which is where
--                     scenariogen controls it anyway.
--
-- The engine ask is filed in .tasks/notes/model-integration.md (2026-08-06):
-- either a `deckHeight` featuredef field that raises the blocking map's
-- sampled Y over the footprint, or extending the FootprintProfile permeability
-- query to features. When that lands, flipping `blocking` here is a one-line
-- change and the models/footprints below need no revision.
--
-- ⚠️ THE ENGINE-SIDE `deckHeight` THAT LANDED (PLAN-maps §2j option C) IS THE
-- SEATING HALF, NOT THE PATHING HALF. `FeatureDef::deckHeight` (resolved from
-- `customparams.deck_top` below) makes a span HOLD the y it was staged at
-- instead of being clamped up to the ground — a level deck. It does NOT raise
-- the blocking map's sampled Y and it does NOT make anything walk over a span.
-- The paragraphs above stand exactly as written: `blocking = false` is still
-- the right half of a binary choice, and flipping it is still gated on the
-- pathing work, which nobody has done. (For a month after §2j option A landed
-- the seating half did not fire on these two defs either, because a truthful
-- `deck_top` of 0 read as "no deck"; that was repaired on 2026-09-20 and is
-- written up in the next section. The pathing half was never contingent on
-- either.)
--
-- ============================================================================
-- THE ORIGIN IS THE DECK (PLAN-maps.md §2j option A, LANDED 2026-08-19)
-- ============================================================================
-- Every span in this file is authored with its ORIGIN AT THE TRAFFICABLE
-- DECK. Measured off the shipped glTF: `ms_road_bridge` runs
-- y = -1.50 .. +3.02 with 28 deck verts at exactly y = 0 spanning
-- x -4.35..+4.35; `ms_rail_bridge` runs y = -3.80 .. +0.35 with its deck slab
-- top at y = 0 and the rail head at +0.35. The substructure — floor beams,
-- bottom chords, pier footings — runs NEGATIVE, which is where a real
-- abutment buries it.
--
-- WHY THE ORIGIN HAD TO MOVE. `CFeature::UpdatePosition` ends every tick with
--
--   Move(UpVector * (max(CGround::GetHeightReal(x, z), pos.y) - pos.y))
--   (Feature.cpp:570)
--
-- — a feature's y is clamped UP to the ground and can never be pushed down.
-- While the origin was the PIER BASE the deck sat `deck_top` above it, so a
-- unit driving at the terrain height under a span was always exactly
-- `deck_top` (1.5 road / 3.8 rail) below the deck it should have been driving
-- on — and **raising the terrain raised the span with it**, so the gap was
-- invariant under every earthwork a map generator can do. That is why it was
-- never a map job, and why §2j's "raised causeway whose top IS the deck" is
-- withdrawn as impossible rather than merely expensive. With y = 0 at the
-- deck, a span standing on ground `g` decks at `g`, exactly where the road it
-- meets is: on dry ground the clamp now lands the deck FLUSH.
--
-- ⚠️ THIS DOES NOT FIX A SPAN OVER WATER, and §2j said so before it landed.
-- A chain staged at the waterline (y = 0, which is what scenariogen does —
-- see game_scenario.lua's stageFeatures) now decks AT the waterline rather
-- than 1.5 above it, while units wade the bed below: the gap becomes the
-- ford's depth instead of a flat 1.5. Better where the ford is shallower than
-- 1.5, worse where it is deeper. The water case belongs to C (`deckHeight`)
-- plus a scenario that stages at a chosen deck level — not to A.
--
-- The user ruled A + C on 2026-08-19; both have now landed (C first, §2p).
--
-- 🔻 `deck_top` IS NOW ZERO, AND THAT IS THE TRUTHFUL VALUE — the deck is the
-- origin, so the offset between them is nothing. It is still published, per
-- def and not in the shared `span()` posture, because it is the number that
-- says WHERE THE DECK IS relative to the model, and a consumer that stops
-- finding it cannot tell "no deck" from "deck at 0" (ms_defs.feature_deck_top
-- raises rather than defaulting, for exactly that reason).
--
-- ✅ THE ENGINE NOW AGREES — REPAIRED 2026-09-20. For a month it did not, and
-- the shape of that bug is worth keeping: `FeatureSeating::IsSeated` was
-- `deckHeight > 0.0f`, so a truthful `deck_top = 0` read as "no deck declared"
-- and NEITHER of these two defs was seated. The encoding predated A — it was
-- written when a deck could not be at the origin — and A silently cancelled C.
-- It had no live effect (scenariogen stages every chain at y = 0 with every
-- span centre over water via `chain_is_afloat_at`, and at pos.y <= 0 the engine
-- sets the INWATER bit, SolidObject.cpp:48, so `floating = true` below zeroes
-- gravity and the chain holds y = 0 regardless — exactly the 0/0/0/0 §M3
-- measured BEFORE seating existed), but it lost the case seating was built for:
-- a level deck over a DRY ravine.
--
-- The repair split the two facts apart. `FeatureSeating::DeckSpec` carries
-- `declared` and `height` separately; `ResolveDeck` derives `declared` from the
-- PRESENCE of this key, never from its value, which is the same raise-don't-
-- default contract `ms_defs.feature_deck_top` already kept on the Python side.
-- So publishing `deck_top = '0'` below is what SEATS these defs, and deleting
-- the key is what would unseat them. Measured on green_flat_x34_v3 (dry,
-- ground 40.87), three features staged at 80.87: before, ms_road_bridge fell
-- to 40.87 exactly like the deckless ms_train_wreck control; after, it holds
-- 80.87 while the control still falls. `FeatureDefs[d].hasDeck` is published
-- to Lua alongside `deckHeight` for the same reason.
--
-- ============================================================================
-- CHAINING — the acceptance criterion, and why 24.0 exactly is safe
-- ============================================================================
-- Both spans are authored as tileable segments on their local Z (RH, -Z
-- forward). Measured off the shipped glTF:
--
--   ms_road_bridge   Z -12.00 .. +12.00   exactly 24.00 m
--   ms_rail_bridge   Z -12.041 .. +12.041      24.08 m
--
-- Chain pitch is therefore exactly **24.0** for both, published as
-- `customparams.chain_pitch` so placement code has one source of truth
-- (game_scenario.lua's `world.features` `chain` field reads it).
--
-- The rail span's extra 41 mm per end is 16 vertices at y=-0.894, x=+-1.74..1.86
-- — the rail heads, overhanging on purpose so a chained line of segments reads
-- as continuous rail. At a 24 m pitch neighbouring rail heads interpenetrate by
-- 82 mm: solid inside solid, no coplanar pair, no z-fighting.
--
-- The road span carries 14 deck cross-section cap triangles on each end plane.
-- Chained at 24 m, segment A's +Z caps and segment B's -Z caps are coincident —
-- but ANTIPARALLEL, and both materials are single-sided (`doubleSided` is unset
-- in the glTF). Whichever one faces the camera is the only one drawn, and it
-- sits inside the neighbour's solid, so there is no coincident same-facing pair
-- to fight. `tools/forge/samples/ms_road_bridge/ms_road_bridge_layout.py:5`
-- states the intent: "deck runs exactly z = -12..+12 at full width at both ends
-- so segments chain seamlessly".
--
-- Do NOT "fix" the seam by nudging the pitch off 24. A 23.9 m pitch would
-- overlap the decks and put the two deck TOP surfaces — which are coplanar,
-- same-facing and textured — into genuine z-fighting. The current geometry is
-- only safe because the coincident faces point away from each other.

--- Shared span posture. Cosmetic scenery: no blocking (see the header), no
--- selecting, and indestructible so a stray shell cannot delete a span while
--- the terrain crossing it explains stays right where it was.
local function span(t)
    t.blocking       = false
    t.noselect       = true
    t.indestructible = true            -- implies reclaimable = false; correct, a bridge is not scrap
    t.flammable      = false
    t.upright        = true
    -- FLOATING IS REDUNDANT FOR A SEATED DEF, AND KEPT ANYWAY. A seated
    -- feature never runs the gravity term `floating` zeroes, so with the
    -- seating repair of 2026-09-20 (see the header) every def in this file is
    -- seated and `floating` changes nothing for them. It is kept because it
    -- costs nothing, because it is the correct posture for a span regardless,
    -- and because it is the belt to seating's braces: between option A landing
    -- and the repair, `floating` was the ONLY thing holding a chain level over
    -- water, and that is not a hole worth reopening for a one-line saving.
    --
    -- Features are not fixed in the air: CFeature::UpdatePosition applies
    -- gravity every tick and then clamps to
    -- `max(CGround::GetHeightReal(x, z), pos.y)` (Feature.cpp:565-571), so the
    -- y a scenario passes to Spring.CreateFeature is a SPAWN height that the
    -- feature immediately settles out of. Live-measured 2026-08-06: four road
    -- spans placed at y = 0 over a skerry_reach channel came to rest at
    -- -31.0 / -34.5 / -45.9 / -57.6 — a staircase down the seabed, which is
    -- exactly what a bridge must not be.
    --
    -- `floating` zeroes the gravity term while the feature is in water
    -- (Feature.cpp:533, `gravAccel * (1 - def->floating)`), so a span spawned
    -- at the waterline STAYS at the waterline and the chain stays level. That
    -- covers the case §M4 actually wants ("bridges over water gaps —
    -- scorched_crossing is the natural first map").
    --
    -- What floating did NOT fix: a span over a DRY ravine fell to the riverbed,
    -- because the ground clamp was unconditional. Nor does terrain shaping fix
    -- it — see the deck-gap header above: the clamp raises the span with the
    -- ground, so an earthwork moves the deck and the road surface by the same
    -- amount and the gap never closes. **`deck_top` seating fixed that** — a
    -- span with a declared deck holds its staged y on dry ground too — and
    -- §2j option A then closed the gap itself by putting y = 0 ON the deck,
    -- at the cost of the declaration the seating test recognises. What A buys
    -- unconditionally is the case seating never could: a MAP-placed span,
    -- which the featureplacer spawns at `CGround::GetHeightReal` with no y at
    -- all, now decks AT that ground instead of 1.5 above it. Laying a LEVEL
    -- deck across uneven ground is still the scenario path's alone
    -- (game_scenario.lua's stageFeatures), and still wants seating back.
    t.floating       = true
    t.smokeTime      = 0
    t.metal          = 0
    t.energy         = 0
    t.customparams   = t.customparams or {}
    -- Defaults, not overrides: the ancient span below publishes its own kind
    -- and its own (36 m) pitch, and a posture helper must not silently undo a
    -- def's measured number.
    t.customparams.ms_feature_kind = t.customparams.ms_feature_kind or 'bridge'
    t.customparams.chain_axis  = 'z'   -- tiles along local Z (RH, -Z forward)
    t.customparams.chain_pitch = t.customparams.chain_pitch or '24'  -- metres between segment centres; measured, exact
    t.customparams.cosmetic_span = '1' -- non-blocking pending the deck-pathing engine ask
    t.customparams.generator = 'Claude Fable 5 (tools/forge)'
    return t
end

return {
    -- 8.9 m wide, 4.5 m tall, 24.0 m per segment. Riveted steel through-truss
    -- with a cracked concrete deck and pier footings. The road crossing:
    -- scorched_crossing is the natural first map for it (§M4).
    ms_road_bridge = span{
        description = 'Road bridge span — riveted through-truss, 24 m segment',
        object      = 'ms_road_bridge',
        footprintx  = 4, footprintz = 12,     -- 8 x 24 m
        health      = 6000,                   -- inert while indestructible; sized for the flip
        mass        = 9000,
        customparams = {
            -- Roadway slab surface above the model's y = 0 — ZERO, because
            -- §2j option A made the roadway slab BE y = 0. Measured off the
            -- shipped mesh: 28 verts at y = 0.000 spanning x -4.35..+4.35,
            -- kerb tops 0.22 above it, model floor at -1.50 (was 1.500 above
            -- a pier-base origin running 0.00..4.52). Read by the ENGINE
            -- (FeatureDef::deck) as well as by content: PRESENCE of this key
            -- is what seats the def, and zero is simply the offset. See the
            -- header — that has only been true since 2026-09-20.
            deck_top = '0',
        },
    },

    -- 4.4 m wide, 4.2 m tall, 24.0 m per segment (rail heads to 24.08). Deck
    -- truss carrying track on top, pier footings below. Pairs with fable_train
    -- routes and with ms_train_wreck when the line is meant to read as cut.
    ms_rail_bridge = span{
        description = 'Rail bridge span — deck truss, track on top, 24 m segment',
        object      = 'ms_rail_bridge',
        footprintx  = 2, footprintz = 12,     -- 4 x 24 m
        health      = 5000,
        mass        = 7000,
        customparams = {
            -- Deck slab top, NOT the rail head (now +0.35): the deck is what
            -- a vehicle would stand on and what an abutment has to meet, and
            -- it is the surface §2j option A put at y = 0. Model floor at
            -- -3.80 (was 3.80 above a pier-base origin running 0.00..4.15).
            -- Presence seats it; the 0 is just the offset. See the header.
            deck_top = '0',
        },
    },

    -- ------------------------------------------------------------------
    -- Batch-04 ancient span (units-assets review finding 9, 2026-09-17)
    -- ------------------------------------------------------------------
    -- 12.9 m wide, 13.4 m tall, 36.0 m per segment. One impossible shallow
    -- monolithic arc with no mid-span supports, a seamless deck carrying an
    -- active cyan guide-channel, a perfect floating alloy ring threaded
    -- around mid-span, half-plinth footings at both tile ends so segments
    -- chain. The ancient counterpart of the two steel spans above, and the
    -- same posture: cosmetic, non-blocking, unselectable, permanent, floating.
    --
    -- ⚠️ THE ORIGIN IS THE FOOTINGS, NOT THE DECK. This model predates §2j
    -- option A and was never re-authored: measured off the shipped glTF the
    -- footings sit at y = 0 and the trafficable deck is a 136-vertex plateau
    -- at y = +7.0 m (56 elmos) across the mid-span. So unlike the two spans
    -- above, its `deck_top` is POSITIVE. That difference used to decide
    -- whether the engine seated a def at all; since the 2026-09-20 repair it
    -- decides only WHERE the deck is, and this span is seated for the same
    -- reason the two above are — it publishes the key. (It was, in fact, the
    -- only span still seated during the month the sign test was live.) Over
    -- water that changes nothing (`floating` already holds it). On dry ground
    -- it is the case seating was built for: stage it at grade and the deck
    -- reads 7 m above the valley floor, footings on the ground. Value is in
    -- engine world units (elmos), which is what FeatureSeating::
    -- ResolveDeckHeight consumes.
    --
    -- NOT placed automatically. BRIDGE_SPANS names it under "ancient" the way
    -- the rail span is named under "rail": available to hand-authored
    -- scenarios, not chosen by scenariogen's crossing placer, which lays road.
    ms_anc_bridge_span = span{
        description = 'Ancient bridge span — self-supporting monolithic arc, 36 m segment',
        object      = 'ms_anc_bridge_span',
        footprintx  = 6, footprintz = 18,     -- 12 x 36 m
        health      = 30000,
        mass        = 40000,
        customparams = {
            ms_feature_kind = 'ancient',      -- overrides span()'s 'bridge': it is both, and the relic reading wins for consumers that filter on kind
            relic_kind  = 'span',
            chain_pitch = '36',   -- metres between segment centres; measured, exact (overrides span()'s 24)
            deck_top    = '56',   -- elmos: the deck plateau at +7.0 m over a footings origin (seated, like every def that publishes the key)
        },
    },
}
